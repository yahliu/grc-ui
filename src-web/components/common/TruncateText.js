/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
/* Copyright (c) 2020 Red Hat, Inc. */
/* Copyright Contributors to the Open Cluster Management project */


import React from 'react'
import { Tooltip } from '@patternfly/react-core'
import PropTypes from 'prop-types'
import truncate from '../../utils/truncate-middle'

class TruncateText extends React.PureComponent {
  static propTypes = {
    maxCharacters: PropTypes.number,
    maxWidth: PropTypes.string,
    position: PropTypes.string,
    text: PropTypes.oneOfType([PropTypes.array, PropTypes.string]).isRequired,
    textEnd: PropTypes.string
  }

  static defaultProps ={
    maxCharacters: 35,
  }

  render() {
    const {text, maxCharacters, maxWidth, position, textEnd} = this.props
    const postfix = textEnd ? textEnd : ''
    if (!text) {
      return ''
    }
    else if (typeof text !== 'string') {
      return `${text.toString()}${postfix}`
    }
    else if (text.length <= maxCharacters){
      return `${text}${postfix}`
    }

    return <Tooltip
      maxWidth = {maxWidth ? maxWidth : '80rem'}
      enableFlip = {false}
      position = {position ? position : 'top'}
      content = {<div>{text}</div>}
    >
      <span className='textWithTruncation'>{`${truncate(text, maxCharacters)}${postfix}`}</span>
    </Tooltip>
  }
}

export default TruncateText
