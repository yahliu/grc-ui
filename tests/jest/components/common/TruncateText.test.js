/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
/* Copyright (c) 2020 Red Hat, Inc. */
/* Copyright Contributors to the Open Cluster Management project */

'use strict'

import React from 'react'
import TruncateText from '../../../../src-web/components/common/TruncateText'
import { shallow } from 'enzyme'

describe('TruncateText', () => {
  it('renders as expected for truncated text', () => {
    const props = { maxCharacters: 7, text: 'TruncateText'}
    const component = shallow(
      <TruncateText  {...props} />
    )
    expect(component.instance()).toMatchSnapshot()
  })

  it('renders as expected for truncated text', () => {
    const props = {
      maxCharacters: 5,
      text: 'TruncateText',
      position: 'bottom',
      textEnd: '.',
      maxWidth: '30rem'
    }
    const component = shallow(
      <TruncateText  {...props} />
    )
    expect(component.instance()).toMatchSnapshot()
  })

  it('renders as expected for non truncated text', () => {
    const props = { maxCharacters: 25, text: 'NonTruncateText'}
    const component = shallow(
      <TruncateText  {...props} />
    )
    expect(component.instance()).toMatchSnapshot()
  })

  it('renders as expected for null', () => {
    const props = { maxCharacters: 25, text: ''}
    const component = shallow(
      <TruncateText  {...props} />
    )
    expect(component.instance()).toMatchSnapshot()
  })

  it('renders as expected for array', () => {
    const props = { maxCharacters: 25, text: [1,2,3]}
    const component = shallow(
      <TruncateText  {...props} />
    )
    expect(component.instance()).toMatchSnapshot()
  })
})
