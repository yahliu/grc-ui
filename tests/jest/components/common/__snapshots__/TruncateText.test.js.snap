// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TruncateText renders as expected for array 1`] = `
TruncateText {
  "context": Object {},
  "props": Object {
    "maxCharacters": 25,
    "text": Array [
      1,
      2,
      3,
    ],
  },
  "refs": Object {},
  "setState": [Function],
  "state": null,
  "updater": Updater {
    "_callbacks": Array [],
    "_renderer": ReactShallowRenderer {
      "_context": Object {},
      "_didScheduleRenderPhaseUpdate": false,
      "_dispatcher": Object {
        "readContext": [Function],
        "useCallback": [Function],
        "useContext": [Function],
        "useDebugValue": [Function],
        "useDeferredValue": [Function],
        "useEffect": [Function],
        "useImperativeHandle": [Function],
        "useLayoutEffect": [Function],
        "useMemo": [Function],
        "useReducer": [Function],
        "useRef": [Function],
        "useResponder": [Function],
        "useState": [Function],
        "useTransition": [Function],
      },
      "_element": <TruncateText
        maxCharacters={25}
        text={
          Array [
            1,
            2,
            3,
          ]
        }
      />,
      "_firstWorkInProgressHook": null,
      "_forcedUpdate": false,
      "_instance": [Circular],
      "_isReRender": false,
      "_newState": null,
      "_numberOfReRenders": 0,
      "_renderPhaseUpdates": null,
      "_rendered": "1,2,3",
      "_rendering": false,
      "_updater": [Circular],
      "_workInProgressHook": null,
    },
  },
}
`;

exports[`TruncateText renders as expected for non truncated text 1`] = `
TruncateText {
  "context": Object {},
  "props": Object {
    "maxCharacters": 25,
    "text": "NonTruncateText",
  },
  "refs": Object {},
  "setState": [Function],
  "state": null,
  "updater": Updater {
    "_callbacks": Array [],
    "_renderer": ReactShallowRenderer {
      "_context": Object {},
      "_didScheduleRenderPhaseUpdate": false,
      "_dispatcher": Object {
        "readContext": [Function],
        "useCallback": [Function],
        "useContext": [Function],
        "useDebugValue": [Function],
        "useDeferredValue": [Function],
        "useEffect": [Function],
        "useImperativeHandle": [Function],
        "useLayoutEffect": [Function],
        "useMemo": [Function],
        "useReducer": [Function],
        "useRef": [Function],
        "useResponder": [Function],
        "useState": [Function],
        "useTransition": [Function],
      },
      "_element": <TruncateText
        maxCharacters={25}
        text="NonTruncateText"
      />,
      "_firstWorkInProgressHook": null,
      "_forcedUpdate": false,
      "_instance": [Circular],
      "_isReRender": false,
      "_newState": null,
      "_numberOfReRenders": 0,
      "_renderPhaseUpdates": null,
      "_rendered": "NonTruncateText",
      "_rendering": false,
      "_updater": [Circular],
      "_workInProgressHook": null,
    },
  },
}
`;

exports[`TruncateText renders as expected for null 1`] = `
TruncateText {
  "context": Object {},
  "props": Object {
    "maxCharacters": 25,
    "text": "",
  },
  "refs": Object {},
  "setState": [Function],
  "state": null,
  "updater": Updater {
    "_callbacks": Array [],
    "_renderer": ReactShallowRenderer {
      "_context": Object {},
      "_didScheduleRenderPhaseUpdate": false,
      "_dispatcher": Object {
        "readContext": [Function],
        "useCallback": [Function],
        "useContext": [Function],
        "useDebugValue": [Function],
        "useDeferredValue": [Function],
        "useEffect": [Function],
        "useImperativeHandle": [Function],
        "useLayoutEffect": [Function],
        "useMemo": [Function],
        "useReducer": [Function],
        "useRef": [Function],
        "useResponder": [Function],
        "useState": [Function],
        "useTransition": [Function],
      },
      "_element": <TruncateText
        maxCharacters={25}
        text=""
      />,
      "_firstWorkInProgressHook": null,
      "_forcedUpdate": false,
      "_instance": [Circular],
      "_isReRender": false,
      "_newState": null,
      "_numberOfReRenders": 0,
      "_renderPhaseUpdates": null,
      "_rendered": "",
      "_rendering": false,
      "_updater": [Circular],
      "_workInProgressHook": null,
    },
  },
}
`;

exports[`TruncateText renders as expected for truncated text 1`] = `
TruncateText {
  "context": Object {},
  "props": Object {
    "maxCharacters": 7,
    "text": "TruncateText",
  },
  "refs": Object {},
  "setState": [Function],
  "state": null,
  "updater": Updater {
    "_callbacks": Array [],
    "_renderer": ReactShallowRenderer {
      "_context": Object {},
      "_didScheduleRenderPhaseUpdate": false,
      "_dispatcher": Object {
        "readContext": [Function],
        "useCallback": [Function],
        "useContext": [Function],
        "useDebugValue": [Function],
        "useDeferredValue": [Function],
        "useEffect": [Function],
        "useImperativeHandle": [Function],
        "useLayoutEffect": [Function],
        "useMemo": [Function],
        "useReducer": [Function],
        "useRef": [Function],
        "useResponder": [Function],
        "useState": [Function],
        "useTransition": [Function],
      },
      "_element": <TruncateText
        maxCharacters={7}
        text="TruncateText"
      />,
      "_firstWorkInProgressHook": null,
      "_forcedUpdate": false,
      "_instance": [Circular],
      "_isReRender": false,
      "_newState": null,
      "_numberOfReRenders": 0,
      "_renderPhaseUpdates": null,
      "_rendered": <mockTooltip
        content={
          <div>
            TruncateText
          </div>
        }
        enableFlip={false}
        maxWidth="80rem"
        position="top"
      >
        <span
          className="textWithTruncation"
        >
          Tr...xt
        </span>
      </mockTooltip>,
      "_rendering": false,
      "_updater": [Circular],
      "_workInProgressHook": null,
    },
  },
}
`;

exports[`TruncateText renders as expected for truncated text 2`] = `
TruncateText {
  "context": Object {},
  "props": Object {
    "maxCharacters": 5,
    "maxWidth": "30rem",
    "position": "bottom",
    "text": "TruncateText",
    "textEnd": ".",
  },
  "refs": Object {},
  "setState": [Function],
  "state": null,
  "updater": Updater {
    "_callbacks": Array [],
    "_renderer": ReactShallowRenderer {
      "_context": Object {},
      "_didScheduleRenderPhaseUpdate": false,
      "_dispatcher": Object {
        "readContext": [Function],
        "useCallback": [Function],
        "useContext": [Function],
        "useDebugValue": [Function],
        "useDeferredValue": [Function],
        "useEffect": [Function],
        "useImperativeHandle": [Function],
        "useLayoutEffect": [Function],
        "useMemo": [Function],
        "useReducer": [Function],
        "useRef": [Function],
        "useResponder": [Function],
        "useState": [Function],
        "useTransition": [Function],
      },
      "_element": <TruncateText
        maxCharacters={5}
        maxWidth="30rem"
        position="bottom"
        text="TruncateText"
        textEnd="."
      />,
      "_firstWorkInProgressHook": null,
      "_forcedUpdate": false,
      "_instance": [Circular],
      "_isReRender": false,
      "_newState": null,
      "_numberOfReRenders": 0,
      "_renderPhaseUpdates": null,
      "_rendered": <mockTooltip
        content={
          <div>
            TruncateText
          </div>
        }
        enableFlip={false}
        maxWidth="30rem"
        position="bottom"
      >
        <span
          className="textWithTruncation"
        >
          T...t.
        </span>
      </mockTooltip>,
      "_rendering": false,
      "_updater": [Circular],
      "_workInProgressHook": null,
    },
  },
}
`;
